core/domain/exp_jobs_one_off.py

```diff
@@ -164,120 +248,163 @@ class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
-                                rule_input = re.sub(
-                                    r'%s(\^\d)\((.)\)' % trig_fn,
-                                    r'%s(\2)\1' % trig_fn, rule_input)
-
-                            # Adding parens to trig functions that don't have
-                            # any. For eg. 'cosA' -> 'cos(A)'.
-                            for trig_fn in trig_fns:
-                                rule_input = re.sub(
-                                    r'%s(?!\()(.)' % trig_fn,
-                                    r'%s(\1)' % trig_fn, rule_input)
-
-                            # The pylatexenc lib outputs the unicode values of
-                            # special characters like sqrt and pi, which is why
-                            # they need to be replaced with their corresponding
-                            # text values before performing validation.
-                            for unicode_char, text in (
-                                    unicode_to_text_mapping.items()):
-                                rule_input = rule_input.replace(
-                                    unicode_char, text)
-
-                            # Replacing trig functions that have format which is
-                            # incompatible with the validations.
-                            for invalid_trig_fn, valid_trig_fn in (
-                                    inverse_trig_fns_mapping.items()):
-                                rule_input = rule_input.replace(
-                                    invalid_trig_fn, valid_trig_fn)
-
-                            validity = 'Invalid'
-                            if is_valid_math_expression(rule_input):
-                                validity = 'Valid Expression'
+                            rule_input = _clean_math_expression(rule_input)
+
+                            type_of_input = _TYPE_INVALID_EXPRESSION
+                            if is_valid_algebraic_expression(rule_input):
+                                type_of_input = _TYPE_VALID_ALGEBRAIC_EXPRESSION
+                            elif is_valid_numeric_expression(rule_input):
+                                type_of_input = _TYPE_VALID_NUMERIC_EXPRESSION
                             elif is_valid_math_equation(rule_input):
-                                validity = 'Valid Equation'
+                                type_of_input = _TYPE_VALID_MATH_EQUATION
+
+                            types_of_inputs.add(type_of_input)
+
+                            output_values = '%s %s: %s' % (
+                                item.id, state_name, rule_input)
+
+                            yield (type_of_input, output_values.encode('utf-8'))
+
+                    if _TYPE_INVALID_EXPRESSION in types_of_inputs:
+                        yield ('ERROR', (
+                            'There are some invalid inputs that need to be '
+                            'resolved before running the upgrade job.'))
```

Comment:
Which ones? This message doesn't help at all with debugging.

Reply:
The invalid inputs are displayed in the output with the key 'Invalid'. I've added this in the error message as well to avoid confusion.